http-clients/apache-async-client/src/main/java/software/amazon/awssdk/http/apache/async/ProxyConfiguration.java

```diff
@@ -0,0 +1,404 @@
+     *
+     * @see Builder#preemptiveBasicAuthenticationEnabled(Boolean)
+     */
+    public Boolean preemptiveBasicAuthenticationEnabled() {
+        return preemptiveBasicAuthenticationEnabled;
+    }
+
+    @Override
+    public Builder toBuilder() {
+        return builder()
+            .endpoint(endpoint)
+            .username(username)
+            .password(password)
+            .ntlmDomain(ntlmDomain)
+            .ntlmWorkstation(ntlmWorkstation)
+            .nonProxyHosts(nonProxyHosts)
+            .preemptiveBasicAuthenticationEnabled(preemptiveBasicAuthenticationEnabled);
+    }
+
+    /**
+     * Create a {@link Builder}, used to create a {@link ProxyConfiguration}.
+     */
+    public static Builder builder() {
+        return new DefaultClientProxyConfigurationBuilder();
+    }
+
+    @Override
+    public String toString() {
+        return ToString.builder("ProxyConfiguration")
+                       .add("endpoint", endpoint)
+                       .add("username", username)
+                       .add("ntlmDomain", ntlmDomain)
+                       .add("ntlmWorkstation", ntlmWorkstation)
+                       .add("nonProxyHosts", nonProxyHosts)
+                       .add("preemptiveBasicAuthenticationEnabled", preemptiveBasicAuthenticationEnabled)
+                       .build();
+    }
+
+
+    private String resolveHost() {
+        return endpoint != null ? endpoint.getHost()
+                                : resolveValue(null, ProxySystemSetting.PROXY_HOST);
+    }
+
+    private int resolvePort() {
+        int port = 0;
+
+        if (endpoint != null) {
+            port = endpoint.getPort();
+        } else if (useSystemPropertyValues) {
+            port = ProxySystemSetting.PROXY_PORT.getStringValue()
+                                                .map(Integer::parseInt)
+                                                .orElse(0);
+        }
+
+        return port;
+    }
+
+    public String resolveScheme() {
+        return endpoint != null ? endpoint.getScheme() : null;
+    }
+
+    /**
+     * Uses the configuration options, system setting property and returns the final value of the given member.
+     */
+    private String resolveValue(String value, ProxySystemSetting systemSetting) {
+        return ((value == null) && useSystemPropertyValues) ? systemSetting.getStringValue().orElse(null)
+                                                            : value;
+    }
+
+    /**
+     * Returns the Java system property for nonProxyHosts as set of Strings.
+     * See http://docs.oracle.com/javase/7/docs/api/java/net/doc-files/net-properties.html.
+     */
+    private Set<String> parseNonProxyHostsProperty() {
+        String nonProxyHosts = ProxySystemSetting.NON_PROXY_HOSTS.getStringValue().orElse(null);
+
+        if (!StringUtils.isEmpty(nonProxyHosts)) {
+            return Arrays.stream(nonProxyHosts.split("\\|"))
+                         .map(String::toLowerCase)
```

Comment:
The  `toLowerCase()` is locale-specific and might introduce subtile bugs. Use `ROOT` or `ENGLISH`.